### PR TITLE
DAOS-14018 pool: dup operation detection metadata layout

### PR DESCRIPTION
**Pull request by kccain**
This is the first of multiple patches that will implement
a duplicate (retry) RPC/operation detection feature for pool/container
service metadata RPCs.

With this change, a new pool/container service root-level KVS (svc_ops)
is added so that (eventually, after subsequent patches) a recent
history of client opss handled can be kept. And so that duplicate ops
can be detected and handled appropriately. A boolean (svc_ops_enabled)
is also added to the pool rdb.

For old layout pools, or those with the prior rdb default size 128 MiB,
svc_ops_enabled will be false, since the amount of history to be kept
in this design is too large for that rdb size. Related to this, the
default value of DAOS_MD_CAP is increased to 1024 MiB so that new
pools created will have sufficient rdb capacity. A debug log message
is shown during step up to reveal (especially for pool start/create)
if the feature is enabled.

Because of the layout change,  pool and container upgrade logic
is also changed to enable migration from older layouts to this latest
version (e.g., DAOS v2.4 pools at global version 2 -> 3).

Also, some comments and minor logic changes in the container
create/destroy handling execution flows are added, mostly as a TODO,
and to show how duplicate ops scenarios can/will be handled.

Finally, an independent enhancement to pool service distributed start
is included in this patch. When pool service membership is changed
to add new replicas, the existing rdb size is provided as the size
argument, rather than rely on the DAOS_MD_CAP environment variable
value that of course could have changed between server/engine
starts with an administrator's edits of the daos_server.yml file.

Features: pool container
Allow-unstable-test: true

Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by kccain** [comment]
This PR has undergone extensive testing, with some known failures documented below.  Moving this PR to ready for review state. @liw and @wangshilong could you take a look please?

Testing results, all errors seem to be known existing issues.

1. Build 14 FHM(verbs provider) daos_test/suite.py passed, but had an error parsing the log files afterward. successful run https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13078/14/artifact/Functional%20Hardware%20Medium%20Verbs%20Provider/daos_test/suite.py/results.json but the log parsing error may likely be an instance of https://daosio.atlassian.net/browse/DAOS-14052

2. Build 14 FHM failure https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13078/14/artifact/Functional%20Hardware%20Medium/container/fill_destroy_loop.py/results.json likely an instance of https://daosio.atlassian.net/browse/DAOS-14141

3. Build 14 FHM https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13078/14/artifact/Functional%20Hardware%20Medium/container/per_server_fault_domain.py/ seems to be an instance of https://daosio.atlassian.net/browse/DAOS-13985

4. Build 13 Functional EL8 https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13078/13/artifact/Functional%20on%20EL%208/pool/pda.py/  failure seems to be an instance of https://daosio.atlassian.net/browse/DAOS-14137 

5. Build 11 Functional EL8 https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13078/11/artifact/Functional%20on%20EL%208/pool/create_all_vm.py/job.log looks like either https://daosio.atlassian.net/browse/SRE-471 or something very similar to https://daosio.atlassian.net/browse/DAOS-14228

File: src/container/srv_container.c
Comment (liw) [inline src/container/srv_container.c:5171]: [Not requesting changes] There's `DAOS_UUID_STR_SIZE`.

File: src/container/srv_container.c
Comment (liw) [inline src/container/srv_container.c:5219]: [Just a comment] If the request is a dup, I imagine we will directly generate a reply based on the info in the `ops` KVS, and not execute into the core request handling logic (i.e., not enter this `switch` statement, and certainly not attempting to look up the container).

One thing might need some thoughts: If the request is not a dup, and the container is not found (`rc == -DER_NONEXIST`), then we need to put `-DER_NONEXIST` into the `ops` record of this operation. Do we need to differentiating errors that are conclusive results of operations with errors such as `-DER_NOMEM`, `-DER_GRPVER`, `-DER_TIMEDOUT`?

File: src/include/daos_srv/pool.h
Comment (liw) [inline src/include/daos_srv/pool.h:174]: [Nit] Strictly speaking, we should name this `ds_pool_svc_op_key` (i.e., with the `ds_pool_` prefix).

File: src/include/daos_srv/pool.h
Comment (liw) [inline src/include/daos_srv/pool.h:180]: [Nit] Ditto for `ds_pool_`.

**Review by liw — APPROVED** [review]

File: src/include/daos_srv/pool.h
Comment (kccain) [reply src/include/daos_srv/pool.h:174]: OK, makes sense. Since it is to be used by both pool and container perhaps I kept it non-specific with that in mind. I can make the adjustment in the next patch.

File: src/container/srv_container.c
Comment (kccain) [reply src/container/srv_container.c:5171]: Sure, will fix that up in a follow-on patch for this work, thanks.

**Comment by kccain** [comment]
While waiting for the second review by @wangshilong I have merged latest master, and arranged for just the per-pr testing to be run (since older builds ran with Features: pool container). If CI testing goes cleanly, then this will avoid a need for a forced landing later when the PR gets to the daos gatekeeper.

**Comment by kccain** [comment]
> Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE. https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13078/15/execution/node/1267/log

The functional tests all passed here, though the cart_logparse.py program spotted an unrelated bug in the engine log contents produced while running daos_test/suite.py. I filed a new ticket for that issue https://daosio.atlassian.net/browse/DAOS-14451

**Comment by kccain** [comment]
Adding @wangdi1 to the code review, since @wangshilong may not be available soon for reviewing this PR

**Review by wangshilong — APPROVED** [review]

**Comment by wangshilong** [comment]
@kccain Sorry for late reply..

**Comment by kccain** [comment]
@daos-stack/daos-gatekeeper please use the commit message text from github, (i.e., the text from the message in commit [https://github.com/daos-stack/daos/pull/13078/commits/52198db2dba2c90a2edeb1ccb402281320dcd932](52198db) when landing.

**Comment by mchaarawi** [comment]
> @daos-stack/daos-gatekeeper please use the commit message text from github, (i.e., the text from the message in commit [https://github.com/daos-stack/daos/pull/13078/commits/52198db2dba2c90a2edeb1ccb402281320dcd932](52198db) when landing.

unfortunately landing Li Wei's PR caused a conflict with this. can you please resolve?

**Comment by liw** [comment]
> unfortunately landing Li Wei's PR caused a conflict with this. can you please resolve?

Merged master to resolve these conflicts in `pool_svc_reconf_ult`.

**Review by liw — APPROVED** [review]

**Comment by kccain** [comment]
> Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE. https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13078/17/execution/node/1268/log

functional test passed, but cart_logparse.py found an issue in the engine log, same issue as seen before and is affecting multiple PRs - https://daosio.atlassian.net/browse/DAOS-14450

**Comment by kccain** [comment]
@daos-stack/daos-gatekeeper conflict in srv_pool.c resolved by @liw and per-pr testing has been re-run with one post-successful-test failure as noted. Can this PR be landed? Thanks!

**Comment by kccain** [comment]
> functional test passed, but cart_logparse.py found an issue in the engine log, same issue as seen before and is affecting multiple PRs - https://daosio.atlassian.net/browse/DAOS-14450

same issue as in build 17 (cart_logparse.py found invalid data in engine log entry). I guess someone must have tried to rerun the test hardware stage to try to get a clean CI run (I didn't restart the testing in this case)?

**Review by mchaarawi — APPROVED** [review]

**Pull request merged**
